report_thesis/src/sections/proposed_approach/optimization_framework.tex

```diff
@@ -0,0 +1,193 @@
+However, depending on the number of models in consideration the number of trials can quickly become computationally prohibitive.
+
+In Lines~\ref{step:sample_hyperparameters} to~\ref{step:instantiate_dim_reduction} we instantiate the model and the preprocessors with the hyperparameters sampled by the sampler.
+The sampling process of these is what is being guided by the objective function based on the metrics returned.
+In our setup, we always require that a scaler is instantiated, to ensure that the data is correctly scaled, however, data transformation and dimensionality reduction are not always beneficial, therefore these can be set to the identity function.
+
+Once the preprocessors are instantiated, we construct a pipeline of these to ensure that the data is processed in the order these are defined, seen in Line~\ref{step:construct_pipeline}.
+The order is important, as the preprocessing steps are to a degree interdependent: scaling ensures that all features are on a similar range, which is ensures that transformations are applied uniformly.
+Similarly, dimensionality reduction techniques typically also produce better results if the data has been scaled.
+However, dimensionality reduction may not always be beneficial or may not always produce better results if the data has been transformed.
+As such, we allow for the optimization framework to optionally use these if they are beneficial.
+
+In Line~\ref{step:get_data} to Line~\ref{step:apply_pipeline_cv} we fetch the data, apply our data partitioning strategy to generate four folds, a training set and a test set, and apply the preprocessing to the folds and the training set.
+The purpose of fetching the data for each trial is to ensure that the data is in its default state for each trial.
+This prevents any form of double preprocessing from occuring and potential issues with the generated folds, training set and test set.
+
+As mentioned in Section~\ref{subsec:validation_testing_procedures} we use both cross validation and a test set to evaluate the model.
+This can be seen in line Line~\ref{step:cross_validate} and Lines~\ref{step:train_model} to~\ref{step:evaluate_model}.
+Its important to note, that in practice, the model $m$ is being reinstantiated in each iteration of the cross-validation and again before the model is trained.
+
+Finally, the pruner is being applied in Line~\ref{step:prune_check}.
+In practice, the pruner is being supplied the \texttt{rmse\_cv} and \texttt{std\_dev\_cv} through the objective function, but effectively this amounts to the pruner receiving these metrics directly, as illustrated.
+If the pruner determines that the current trial is unpromising, it stops the trial early, and the optimization process continues to the next trial.
+
+Once a trial is complete, the metrics are returned in Line~\ref{step:return_metrics} to the objective function, which then determines the next steps in the optimization process.
```

Comment:
And again, this is why I want you to fix the `for .. trial t` stuff in the objective function. How would it know, when it only returns after iterating up to `N` trials?